### PR TITLE
BUG: Series.isin fails or categoricals

### LINKED ISSUE #16639: Series.isin fails (errors) for categoricals
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np
#%%
print(pd.__version__)
vals = np.array([0, 1,2, 0]);
cats = ['a', 'b', 'c'];

DFtrades = pd.DataFrame({'id': pd.Series(pd.Categorical(1).from_codes(vals, cats))});
DFscores = pd.DataFrame({'id': pd.Series(pd.Categorical(1).from_codes(np.array([0, 1]), cats))});

print(DFtrades)
print(DFscores)

select_ids = DFtrades['id'].isin(DFscores['id']);

```
#### Problem description

I get an error in 0.20.1

  File "<ipython-input-17-44ce14b18f50>", line 12, in <module>
    select_ids = DFtrades['id'].isin(DFscores['id']);

  File "C:\Users\alexandre\Anaconda3\lib\site-packages\pandas\core\series.py", line 2555, in isin
    result = algorithms.isin(_values_from_object(self), values)

  File "C:\Users\alexandre\Anaconda3\lib\site-packages\pandas\core\algorithms.py", line 421, in isin
    return f(comps, values)

  File "C:\Users\alexandre\Anaconda3\lib\site-packages\pandas\core\algorithms.py", line 399, in <lambda>
    f = lambda x, y: htable.ismember_object(x, values)

  File "pandas\_libs\hashtable_func_helper.pxi", line 428, in pandas._libs.hashtable.ismember_object (pandas\_libs\hashtable.c:29677)

ValueError: Buffer dtype mismatch, expected 'Python object' but got 'long long'

#### Expected Output
a boolean array (or series?) indicating the third row of DFtrades is not in DFscores but the other three are

for reference, this worked (I did not get an error) in 0.19.(something) 

also this code will work as expected:

`select_ids = DFtrades['id'].isin(DFscores['id'].values);`

#### Output of ``pd.show_versions()``
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en
LOCALE: None.None

pandas: 0.20.1
pytest: 3.1.1
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.11.3
scipy: 0.19.0
xarray: 0.9.5
IPython: 6.1.0
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.2.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.8.0
bs4: 4.6.0
html5lib: 0.999
sqlalchemy: 1.1.10
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by aviolov**
- closes #16639 
 - [1 ] tests added / passed
 - ] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff`` (On Windows, ``git diff upstream/master -u -- "*.py" | flake8 --diff`` might work as an alternative.)
 - [check ] whatsnew entry

### CODE REVIEW

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -3,14 +3,18 @@
...
+from pandas.compat import string_types
+from pandas.compat.numpy import _np_version_under1p10
+from pandas.core import common as com
+
+import numpy as np
+from pandas._libs import algos, lib, hashtable as htable
+from pandas._libs.tslib import iNaT
 from pandas.core.dtypes.cast import maybe_promote
-from pandas.core.dtypes.generic import (
-    ABCSeries, ABCIndex,
```
Comment (jreback) [inline pandas/core/algorithms.py]: is there a reason you moved the imports around?

File: pandas/tests/test_categorical.py [hunk 2]
```diff
@@ -442,6 +442,17 @@ def f():
         if hasattr(np.random, "choice"):
             codes = np.random.choice([0, 1], 5, p=[0.9, 0.1])
             pd.Categorical.from_codes(codes, categories=["train", "test"])
+
```
Comment (jreback) [inline pandas/tests/test_categorical.py]: make this a new test, move it to ``test_algos.py`` (near the other isin tests)

File: pandas/tests/test_categorical.py [hunk 3]
```diff
@@ -442,6 +442,17 @@ def f():
...
             pd.Categorical.from_codes(codes, categories=["train", "test"])
+        
+        # Regression test https://github.com/pandas-dev/pandas/issues/16639
+        vals = np.array([0, 1, 2, 0]);
+        cats = ['a', 'b', 'c'];
+        
+        D = pd.DataFrame({'id': pd.Series(pd.Categorical(1).from_codes(vals, cats))});
+        T = pd.DataFrame({'id': pd.Series(pd.Categorical(1).from_codes(np.array([0, 1]), cats))});
+        
+        select_ids = D['id'].isin(T['id']);
```
Comment (jreback) [inline pandas/tests/test_categorical.py]: use result=
the expected= is below and use
``tm.assert_numpy_array_equal``

File: pandas/tests/test_categorical.py [hunk 4]
```diff
@@ -442,6 +442,17 @@ def f():
         if hasattr(np.random, "choice"):
             codes = np.random.choice([0, 1], 5, p=[0.9, 0.1])
             pd.Categorical.from_codes(codes, categories=["train", "test"])
+        
+        # Regression test https://github.com/pandas-dev/pandas/issues/16639
+        vals = np.array([0, 1, 2, 0]);
+        cats = ['a', 'b', 'c'];
+        
+        D = pd.DataFrame({'id': pd.Series(pd.Categorical(1).from_codes(vals, cats))});
```
Comment (jreback) [inline pandas/tests/test_categorical.py]: you have some linting errors, these lines are too long

File: doc/source/whatsnew/v0.21.0.txt [hunk 5]
```diff
@@ -188,7 +188,7 @@ Numeric
 
 Categorical
 ^^^^^^^^^^^
-
+- Bug in ``Series.isin()`` when called for categoricals (:issue`16639`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: use :func:`Series.isin()`

when called with a categorical

**Review by jreback — CHANGES_REQUESTED** [review]
pretty good. some stylistic comments.

File: pandas/core/algorithms.py (hunk 1 above)
Comment (aviolov) [inline pandas/core/algorithms.py]: NOne, that was accidental, possibly due to Pydev or some other fluke, Ive tried to fix to baseline

**Comment by aviolov** [comment]
I"m not sure I understand the Travis CI build failure. Is there something I need to do/change

**Comment by jreback** [comment]
linting is failing

git diff upstream/master --name-only -- '*.py' | flake8 --diff

**Comment by aviolov** [comment]
Hm... I must be missing something, on my local copy I get 
``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
with no output, which i take to mean that all is well. 
Also 
`` $ git status
On branch bug16639-fix
nothing to commit, working directory clean``

Do I need to make another pull request for my minor fixes to the review

**Comment by jreback** [comment]
> Do I need to make another pull request for my minor fixes to the review

NO

simply update and push

**Comment by aviolov** [comment]
From my local 
```$ git pull origin bug16639-fix 
From http://github.com/aviolov/pandas
 * branch            bug16639-fix -> FETCH_HEAD
Already up-to-date.
 $ git push origin bug16639-fix 
Everything up-to-date
`` 

Do I need to apply the 'force' flag?
`` git push -f ``

**Comment by jreback** [comment]
you don't need to force push unless the existing fails. but you are failing linting.

http://pandas.pydata.org/pandas-docs/stable/contributing.html

**Comment by aviolov** [comment]
I got it, I needed to run the 'comprehensive' flakes command

``` 
git diff master --name-only -- '*.py' | grep 'pandas/' | xargs -r flake8
```

Ok, hopefully fixed those

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Comment by aviolov** [comment]
anytime, thanks for reviewing

**Pull request merged**